Sources/NIOHTTPCompression/HTTPRequestCompressor.swift

```diff
@@ -0,0 +1,163 @@
+                // now we have a body lets add the content-encoding header
+                head.headers.replaceOrAdd(name: "Content-Encoding", value: self.encoding.rawValue)
+                state = .body(head, part)
+            case .body(let head, var body):
+                // we have a head and a body, extend the body with this body part
+                guard case .byteBuffer(var part) = buffer else { preconditionFailure("Expected a ByteBuffer") }
+                body.writeBuffer(&part)
+                state = .body(head, body)
+            case .partialBody(var body):
+                // we have a partial body, extend the partial body with this body part
+                guard case .byteBuffer(var part) = buffer else { preconditionFailure("Expected a ByteBuffer") }
+                body.writeBuffer(&part)
+                state = .partialBody(body)
+            default:
+                preconditionFailure("Unexpected Body")
+            }
+            
+        case .end:
+            switch state {
+            case .head(let head):
+                // only found a head
+                context.write(wrapOutboundOut(.head(head)), promise: nil)
+                context.write(data, promise: pendingWritePromise)
+            case .body(var head, var body):
+                // have head and the whole of the body. Compress body, set content length header and write it all out, including the end
+                let outputBuffer = compressor.compress(inputBuffer: &body, allocator: context.channel.allocator, finalise: true)
+                head.headers.replaceOrAdd(name: "Content-Length", value: outputBuffer.readableBytes.description)
+                context.write(wrapOutboundOut(.head(head)), promise: nil)
+                context.write(wrapOutboundOut(.body(.byteBuffer(outputBuffer))), promise: nil)
+                context.write(data, promise: pendingWritePromise)
+            case .partialBody(var body):
+                // have a section of the body. Compress that section of the body and write it out along with the end
+                let outputBuffer = compressor.compress(inputBuffer: &body, allocator: context.channel.allocator, finalise: true)
+                context.write(wrapOutboundOut(.body(.byteBuffer(outputBuffer))), promise: nil)
+                context.write(data, promise: pendingWritePromise)
+            default:
+                preconditionFailure("Unexpected End")
+            }
+            state = .end
+            compressor.shutdown()
+        }
+    }
+    
+    public func flush(context: ChannelHandlerContext) {
+        switch state {
+        case .head(var head):
+            head.headers.remove(name: "Content-Length")
+            head.headers.replaceOrAdd(name: "Transfer-Encoding", value: "chunked")
+            context.write(wrapOutboundOut(.head(head)), promise: pendingWritePromise)
+            state = .partialBody(context.channel.allocator.buffer(capacity: 0))
```

Comment:
I think there's a bug here.

Consider the scenario where the following three operations occur:

```
.write(.head)
.write(.body)
.flush
```

In this case we enter `func write` and transition `state` to `.head`. Then we enter `func write` again and transition state to `.body` _and add the `Content-Encoding` header_. Then we enter `flush`, remove content-length, add transfer-encoding, and write the data.

Consider now the alternative set of operations:

```
.write(.head)
.flush
.write(.body)
```

Here we'll start the same way: enter `func write` and transition `state` to `.head`. Then we'll enter `func flush`, which will send the head, never adding `Content-Encoding`.

Can we add a test case to cover this?

Reply:
fixed and added tests